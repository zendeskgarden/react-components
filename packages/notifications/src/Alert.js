/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */

import styled from 'styled-components';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import CalloutStyles from '@zendeskgarden/css-callouts';
import { retrieveTheme } from '@zendeskgarden/react-theming';

import Well from './Well';
const COMPONENT_ID = 'notifications.alert';

const VALIDATION = {
  SUCCESS: 'success',
  WARNING: 'warning',
  ERROR: 'error',
  INFO: 'info'
};

/**
 * Supports all `<div>` props
 */
const Alert = styled(Well).attrs(props => ({
  'data-garden-id': COMPONENT_ID,
  'data-garden-version': PACKAGE_VERSION,
  recessed: props.type === VALIDATION.INFO,
  className: classNames(props.className, {
    // Validation types
    [CalloutStyles['c-callout--success']]: props.type === VALIDATION.SUCCESS,
    [CalloutStyles['c-callout--warning']]: props.type === VALIDATION.WARNING,
    [CalloutStyles['c-callout--error']]: props.type === VALIDATION.ERROR,
    [CalloutStyles['c-callout--info']]: props.type === VALIDATION.INFO
  })
}))`
  ${props => retrieveTheme(COMPONENT_ID, props)};
`;

Alert.propTypes = {
  type: PropTypes.oneOf([VALIDATION.SUCCESS, VALIDATION.WARNING, VALIDATION.ERROR, VALIDATION.INFO])
    .isRequired
};

/** @component */
export default Alert;
