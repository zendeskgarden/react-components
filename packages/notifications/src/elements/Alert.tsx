/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */

import React, { HTMLAttributes } from 'react';
import PropTypes from 'prop-types';
import { StyledAlert, StyledIcon } from '../styled';
import { validationIcons, validationHues } from '../utils/icons';
import { ARRAY_VALIDATION_TYPE, VALIDATION_HUE } from '../utils/types';
import { NotificationsContext } from '../utils/useNotificationsContext';

export interface IAlertProps {
  /** Applies alert type styles */
  type: 'success' | 'warning' | 'error' | 'info';
}

/**
 * @extends HTMLAttributes<HTMLDivElement>
 */
export const Alert = React.forwardRef<HTMLDivElement, IAlertProps & HTMLAttributes<HTMLDivElement>>(
  (props, ref) => {
    const hue = validationHues[props.type];
    const Icon = validationIcons[props.type] as any;

    return (
      <NotificationsContext.Provider value={hue as VALIDATION_HUE}>
        <StyledAlert ref={ref} hue={hue} {...props}>
          <StyledIcon hue={hue}>
            <Icon />
          </StyledIcon>
          {props.children}
        </StyledAlert>
      </NotificationsContext.Provider>
    );
  }
);

Alert.displayName = 'Alert';

Alert.propTypes = {
  type: PropTypes.oneOf(ARRAY_VALIDATION_TYPE).isRequired
};
