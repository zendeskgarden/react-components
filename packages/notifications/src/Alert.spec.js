/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */

import React from 'react';
import { render } from 'garden-test-utils';
import Alert from './Alert';

describe('Alert', () => {
  describe('validation', () => {
    it('should render success styling correctly', () => {
      const { container } = render(<Alert type="success" />);

      expect(container.firstChild).toHaveClass('c-callout--success');
    });

    it('should render warning styling correctly', () => {
      const { container } = render(<Alert type="warning" />);

      expect(container.firstChild).toHaveClass('c-callout--warning');
    });

    it('should render error styling correctly', () => {
      const { container } = render(<Alert type="error" />);

      expect(container.firstChild).toHaveClass('c-callout--error');
    });

    it('should render info styling correctly', () => {
      const { container } = render(<Alert type="info" />);

      expect(container.firstChild).toHaveClass('c-callout--info');
    });
  });
});
