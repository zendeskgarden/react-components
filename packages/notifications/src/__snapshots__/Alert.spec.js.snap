// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert validation should render error styling correctly 1`] = `
<Well
  className="c-callout--error "
  data-garden-id="notifications.alert"
  data-garden-version="version"
  type="error"
/>
`;

exports[`Alert validation should render success styling correctly 1`] = `
<Well
  className="c-callout--success "
  data-garden-id="notifications.alert"
  data-garden-version="version"
  type="success"
/>
`;

exports[`Alert validation should render warning styling correctly 1`] = `
<Well
  className="c-callout--warning "
  data-garden-id="notifications.alert"
  data-garden-version="version"
  type="warning"
/>
`;
